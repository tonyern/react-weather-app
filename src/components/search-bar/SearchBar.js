import React, { useEffect, useRef, useState } from "react";
import axios from "axios";
import "./search-bar.css";

const SearchBar = ({ searchProps }) => {
  const openWeatherMapAPI = {
    key: "8713314e929a6ae57041a78e979b402d",
    base: "http://api.openweathermap.org/data/2.5/",
  };

  const [query, setQuery] = useState("");
  const inputRef = useRef();

  useEffect(() => {
    if (inputRef.current) {
      inputRef.current.focus();
    }
  });

  const search = (evt) => {
    if (evt.key === "Enter") {
      axios
        .get(
          `${openWeatherMapAPI.base}weather?q=${query}&units=metric&APPID=${openWeatherMapAPI.key}`
        )
        .then((response) => {
          setQuery("");
          searchProps(response.data);
          console.log(response.data);
        })
        .catch((error) => {
          setQuery("");

          if (error.response) {
            // Error if city name was not found or invalid input.
            console.log(error.response.data);
          } else if (error.request) {
            // Error if no response was received.
            console.log(error.request);
          } else {
            // Other errors.
            console.log(error.message);
          }
        });
    }
  };

  return (
    <div className="search-box">
      <input
        type="text"
        ref={inputRef}
        className="search-bar"
        data-testid="search-input-test"
        placeholder="Search City"
        onChange={(event) => setQuery(event.target.value)}
        value={query}
        onKeyPress={search}
      />
    </div>
  );
};

export default SearchBar;
