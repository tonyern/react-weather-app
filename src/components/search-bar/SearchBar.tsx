import React, { useEffect, useRef, useState } from "react";
import axios, { AxiosResponse } from "axios";
import "./search-bar.css";

const SearchBar = ({ searchProps }): JSX.Element => {
  const openWeatherMapAPI = {
    key: "8713314e929a6ae57041a78e979b402d",
    base: "http://api.openweathermap.org/data/2.5/",
  };

  const [query, setQuery] = useState("");
  const inputRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    if (inputRef.current) {
      inputRef.current.focus();
    }
  });

  const search = (evt: React.KeyboardEvent<HTMLDivElement>): void => {
    if (evt.key === "Enter") {
      axios
        .get(
          `${openWeatherMapAPI.base}weather?q=${query}&units=metric&APPID=${openWeatherMapAPI.key}`
        )
        .then((response: AxiosResponse<any>) => {
          setQuery("");
          searchProps(response.data);
          console.log(response.data);
        })
        .catch((error) => {
          setQuery("");

          if (error.response) {
            // Error if city name was not found or invalid input.
            console.log(error.response.data);
          } else if (error.request) {
            // Error if no response was received.
            console.log(error.request);
          } else {
            // Other errors.
            console.log(error.message);
          }
        });
    }
  };

  return (
    <div className="search-box">
      <input
        type="text"
        ref={inputRef}
        className="search-bar"
        data-testid="search-input-test"
        placeholder="Search City"
        onChange={(event) => setQuery(event.target.value)}
        value={query}
        onKeyPress={(evt: React.KeyboardEvent<HTMLDivElement>) => search(evt)}
      />
    </div>
  );
};

export default SearchBar;
