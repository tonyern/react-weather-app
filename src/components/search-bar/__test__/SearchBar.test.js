import React from "react";
import ReactDOM from "react-dom";
import { render, fireEvent } from "@testing-library/react";
import { renderHook } from "@testing-library/react-hooks";
import "@testing-library/jest-dom/extend-expect";
import SearchBar from "../SearchBar";

/**
 * What needs to be tested:
 *
 * 1) Does it render successfully?
 * 2) Is the hook query set to "" as default?
 * 3) Is the ref hook initialized with no problem?
 * 4) Does the hook useEffect focus the ref hook?
 * 5) Does the search function get API data and catch errors?
 */

it("Search bar renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<SearchBar />, div);
});

it("Renders correctly", () => {
  const { queryByTestId, queryByPlaceholderText } = render(<SearchBar />);
  expect(queryByTestId("search-input-test")).toBeTruthy();
  expect(queryByPlaceholderText("Search City")).toBeTruthy();
});

describe("Input value", () => {
  it("Updates on change", () => {
    const { queryByPlaceholderText } = render(<SearchBar />);

    const searchInput = queryByPlaceholderText("Search City");

    fireEvent.change(searchInput, { target: { value: "dallas" } });

    expect(searchInput.value).toBe("dallas");
  });
});

describe("Search functionality", () => {
  describe("Empty query", () => {
    it("Trigger search function but no data from API", () => {
      const search = jest.fn();

      const { queryByTestId, queryByPlaceholderText } = render(
        <SearchBar search={search} />
      );

      fireEvent.submit(queryByTestId("search-input-test"));
      expect(search).not.toHaveBeenCalled();
    });
  });

  describe("Non-empty query", () => {
    it("Trigger search function but data from API", () => {
      const search = jest.fn();

      const { queryByTestId, queryByPlaceholderText } = render(
        <SearchBar search={search} />
      );

      const searchInput = queryByPlaceholderText("Search City");
      fireEvent.change(searchInput, { target: { value: "dallas" } });

      fireEvent.submit(queryByTestId("search-input-test"));
      expect(search).not.toHaveBeenCalled();
    });
  });
});
