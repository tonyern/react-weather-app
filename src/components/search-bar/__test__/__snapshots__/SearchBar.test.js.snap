// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search bar snapshot test 1`] = `
<div
  className="search-box"
>
  <input
    className="search-bar"
    data-testid="search-input-test"
    onChange={[Function]}
    onKeyPress={[Function]}
    placeholder="Search City"
    type="text"
    value=""
  />
</div>
`;
